File: Algorithms/src/sorting/QuickSort.java
package sorting;

import java.util.Random;

/**
 * project_name:Algorithms
 * package_name:common
 * user: youzipi
 * date: 2015/4/6
 */
public class QuickSort extends Sort {
    public static void sort(Comparable[] a) {
        sort(a, 0, a.length - 1);
    }

    public static void sort(Comparable[] a, int left, int right) {

        if (right <= left)
            return;
        Random random = new Random();
        int mid = random.nextInt(right);//随机指定初始值
        int r = right;
        int l = left;

        exch(a, 0, mid);
        Comparable flag = a[0];
        l++;
        while (true) {
            while (lessorEqual(a[l], flag)) {       //向右寻找比flag大的
                l++;
                if (l == right)
                    break;
            }
            while (lessorEqual(flag, a[r])) {       //向左寻找比flag小的
                r--;
                if (r == 0)
                    break;
            }
            if (l >= r)
                break;
            else
                exch(a, l, r);  //交换a[l],a[r]
        }
        exch(a, r, 0);
        sort(a, 0, l - 1);
        sort(a, l+1, right);
    }

    public static void main(String[] args) {
        Integer[] a = new Integer[]{2,6,8,9,3,5,4,7,0};
        show(a);
        QuickSort.sort(a);
        show(a);
    }
}

File: python/sorting/quicksort.py
# -*- coding: utf-8 -*- 
# user: youzipi
# date: 15-10-14 下午3:09
import random


def quick_sort(a):
    return quick_sort2(a, 0, len(a) - 1)


def quick_sort2(a, low, high):
    length = high - low
    if length <= 1:
        return a

    ran = random.randint(low, high - 1)
    left = low + 1
    right = high
    flag = a[ran]
    a[low], a[ran] = a[ran], a[low]
    while True:
        while a[left] < flag:  # 向右寻找大于flag的数
            left += 1
            if left == high:
                break
        while a[right] > flag:  # 向左寻找小于flag的数
            right -= 1
            if right == low:
                break
        if left < right:
            a[left], a[right] = a[right], a[left]
            print a
        else:
            break
    print a
    a[low], a[right] = a[right], a[low]
    quick_sort2(a, low, right - 1)
    quick_sort2(a, right + 1, high)
    return a


a = [2, 8, 9, 4, 5, 6, 7, 1, 0, 3, ]
print quick_sort(a)
